### ISSUE TITLE
Inherited functions not considered when detecting definitions

### ISSUE THREAD

**Issue by titanicjames** [microsoft/pylance-release#3489]
```
#Base class for all VM instructions.
class vmInstruction():

    #Calls the parent's event handler.
    def raiseEvent(self, eventString, *details):
        if self.parent == None:
            #WTF?
            raise Exception("No parent")
        return self.parent.event(eventString, details)

    #Empty, must be created by the instructions.
    def __call__(self):
        pass

#Tells the parent the code has stopped execution.
class halt(vmInstruction):
    def __call__(self):
    #The next line is marked as an undefined function, despite the function being clearly defined in the base class.
        raiseEvent(self, "halt")
```
        
This might be a bit pedantic, but i'd rather not be bugged with these false positives while working on this.

**Comment by erictraut** [comment]
If you try to run the above code and invoke `halt.__call__`, you'll receive a `NameError: name 'raiseEvent' is not defined` exception. That's because `raiseEvent` is defined in a scope that is not visible at that point in the code.

```python
h = halt()
h() # NameError: name 'raiseEvent' is not defined
```

I think what you intended here is the following:
```python
class halt(vmInstruction):
    def __call__(self):
        self.raiseEvent("halt")
```

This eliminates the `NameError`, but you will now receive a different exception: `AttributeError: 'halt' object has no attribute 'parent'`.

In short, the errors that pylance is reporting are not false positives. They're actual bugs in your code that will result in runtime exceptions.

**Comment by titanicjames** [comment]
Hmm, well that's something i didn't know.
